docs/grammar.md

```diff
@@ -0,0 +1,90 @@
+Syntax
+------
+
+The grammar consists of a sequence of rules of the form:
+```
+rule_name: <expression>
+```
+
+Optionally, a type can be included right after the rule name,
+which specifies the return type of the C function corresponging
+to the rule:
+```
+rule_name[return_type]: <expression>
+```
+If the return type is omitted, then a `void *` is
+returned.
+
+### Grammar Expressions
+
+##### `# comment`
+Python-style comments.
+
+##### `e1 e2`
+Match e1, then match e2.
+
+##### `e1 | e2`
+Match e1 or e2.
+
+A | can be used before the first alternative as well for formatting
+purposes.
```

Comment:
This only applies when the alternatives are formatted on separate lines (like in the return value example below), not in the "short form":
```
rule_name: | alternative1 | alternative2
```
is invalid.